umpost/um2netcdf.py

```diff
@@ -92,46 +102,194 @@ def convert_proleptic(time):
-    elif (lon.points[0] == 0 and grid_type == 'EG') or \
-         (np.allclose(0.5*dlon, lon.points[0]) and grid_type == 'ND'):
-        lon.var_name = 'lon_u'
+        lon_coordinate.var_name = 'lon'
+
+
+def is_lat_river_grid(latitude_points):
+    """
+    Check whether latitude points are on the river routing grid.
+
+    Parameters
+    ----------
+    latitude_points: (array) 1D array of latitude grid points.
+    """
+    return len(latitude_points) == 180
+
+
+def is_lon_river_grid(longitude_points):
+    """
+    Check whether longitude points are on the river routing grid.
+
+    Parameters
+    ----------
+    longitude_points: (array) 1D array of longitude grid points.
+    """
+
+    return len(longitude_points) == 360
+
+
+def is_lat_v_grid(latitude_points, grid_type, dlat):
+    """
+    Check whether latitude points are on the lat_v grid.
+
+    Parameters
+    ----------
+    latitude_points: (array) 1D array of latitude grid points.
+    grid_type: (string) model horizontal grid type.
+    dlat: (float) meridional spacing between latitude grid points.
+    """
+    
+    min_latitude = latitude_points[0]
+    min_latitude_v_ND = -90.+0.5*dlat
+    return (
+        (min_latitude == -90 and grid_type == GRID_END_GAME) or
+        (
+            np.allclose(min_latitude_v_ND, min_latitude) and
+            grid_type == GRID_NEW_DYNAMICS
+        )
+    )
+
+
+def is_lon_u_grid(longitude_points, grid_type, dlon):
+    """
+    Check whether longitude points are on the lon_u grid.
+
+    Parameters
+    ----------
+    longitude: (array) 1D array of longitude grid points.
+    grid_type: (string) model horizontal grid type.
+    dlon: (float) zonal spacing between longitude grid points. 
+    """
+    min_longitude = longitude_points[0]
+    min_longitude_u_ND = 0.5*dlon
+
+    return (
+        (min_longitude == 0 and grid_type == GRID_END_GAME) or
+        (
+            np.allclose(min_longitude_u_ND, min_longitude) and
+            grid_type == GRID_NEW_DYNAMICS
+        )
+    )
+
+
+def add_latlon_coord_bounds(cube_coordinate):
+    """
+    Add bounds to horizontal coordinate (longitude or latitude) if
+    they don't already exist. Edits coordinate in place.
+
+    Parameters
+    ----------
+    cube_coordinate: coordinate object from iris cube.
+    """
+    coordinate_name = cube_coordinate.name()
+    if coordinate_name not in [LON_COORD_NAME, LAT_COORD_NAME]:
+        raise ValueError(
+                f"Wrong coordinate {coordinate_name} supplied. "
+                f"Expected one of {LON_COORD_NAME}, {LAT_COORD_NAME}."
+            )
+
+    if cube_coordinate.has_bounds(): 
+        # Don't change bounds if already exist
+        return
+    
+    if len(cube_coordinate.points) > 1:
+        cube_coordinate.guess_bounds()
     else:
-        lon.var_name = 'lon'
+        # For length 1, assume it's global. guess_bounds doesn't work in this case
+        cube_coordinate.bounds = GLOBAL_COORD_BOUNDS[coordinate_name]
```

Comment:
The logic/flow can be somewhat simplified by inverting the `has_bounds()` check:
```Python
if not cube_coordinate.has_bounds(): 
    if len(cube_coordinate.points) > 1:
        cube_coordinate.guess_bounds()
    else:
        # For length 1, assume it's global. guess_bounds doesn't work in this case
        cube_coordinate.bounds = GLOBAL_COORD_BOUNDS[coordinate_name]
```
I prefer this new code to the original as you've improved overall readability.

Reply:
Great suggestion. I think it helps a lot with readability – I've added this in in the latest commit.